panther_manager/src/safety_manager_node.cpp

```diff
@@ -0,0 +1,291 @@
+  rear_driver_temp_ma_->Roll(driver_state->rear.temperature);
+
+  // to simplify conditions pass only higher temp of motor drivers
+  safety_tree_manager_->GetBlackboard()->set<double>(
+    "driver_temp",
+    std::max({front_driver_temp_ma_->GetAverage(), rear_driver_temp_ma_->GetAverage()}));
+}
+
+void SafetyManagerNode::EStopCB(const BoolMsg::SharedPtr e_stop)
+{
+  safety_tree_manager_->GetBlackboard()->set<bool>("e_stop_state", e_stop->data);
+}
+
+void SafetyManagerNode::IOStateCB(const IOStateMsg::SharedPtr io_state)
+{
+  if (io_state->power_button) {
+    ShutdownRobot("Power button pressed");
+  }
+
+  safety_tree_manager_->GetBlackboard()->set<bool>("aux_state", io_state->aux_power);
+  safety_tree_manager_->GetBlackboard()->set<bool>("fan_state", io_state->fan);
+}
+
+void SafetyManagerNode::SystemStatusCB(const SystemStatusMsg::SharedPtr system_status)
+{
+  cpu_temp_ma_->Roll(system_status->cpu_temp);
+  safety_tree_manager_->GetBlackboard()->set<double>("cpu_temp", cpu_temp_ma_->GetAverage());
+}
+
+void SafetyManagerNode::SafetyTreeTimerCB()
+{
+  if (!SystemReady()) {
+    return;
+  }
+
+  safety_tree_manager_->TickOnce();
+
+  if (safety_tree_manager_->GetTreeStatus() == BT::NodeStatus::FAILURE) {
+    RCLCPP_WARN(this->get_logger(), "Safety behavior tree returned FAILURE status");
+  }
+
+  std::pair<bool, std::string> signal_shutdown;
+  if (safety_tree_manager_->GetBlackboard()->get<std::pair<bool, std::string>>(
+        "signal_shutdown", signal_shutdown)) {
+    if (signal_shutdown.first) {
+      ShutdownRobot(signal_shutdown.second);
+    }
+  }
+}
+
+bool SafetyManagerNode::SystemReady()
+{
+  if (
+    !safety_tree_manager_->GetBlackboard()->getEntry("e_stop_state") ||
+    !safety_tree_manager_->GetBlackboard()->getEntry("battery_status") ||
+    !safety_tree_manager_->GetBlackboard()->getEntry("aux_state") ||
+    !safety_tree_manager_->GetBlackboard()->getEntry("cpu_temp") ||
+    !safety_tree_manager_->GetBlackboard()->getEntry("driver_temp")) {
+    RCLCPP_INFO_THROTTLE(
+      this->get_logger(), *this->get_clock(), 5000,
+      "Waiting for required system messages to arrive");
+    return false;
+  }
+
+  return true;
+}
+
+void SafetyManagerNode::ShutdownRobot(const std::string & reason)
+{
+  RCLCPP_WARN(this->get_logger(), "Soft shutdown initialized. %s", reason.c_str());
+  safety_tree_timer_->cancel();
+  safety_tree_manager_->HaltTree();
+
+  // tick shutdown tree
+  auto status = BT::NodeStatus::RUNNING;
+  auto start_time = this->get_clock()->now();
+  rclcpp::Rate rate(30.0);  // 30 Hz
+  while (rclcpp::ok() && status == BT::NodeStatus::RUNNING) {
+    shutdown_tree_manager_->TickOnce();
+    status = shutdown_tree_manager_->GetTreeStatus();
+    rate.sleep();
+  }
+
+  if (status == BT::NodeStatus::FAILURE) {
+    RCLCPP_WARN(
+      this->get_logger(),
+      "Shutdown behavior tree returned FAILURE status, robot may not be shutdown correctly");
+  }
+
+  rclcpp::shutdown();
+}
```

Comment:
So, there are hardly any failure scenarios. There are only 2 bt nodes that will be triggered in default shutdown tree, and both of them were disabled in this [pr](https://github.com/husarion/panther_ros/pull/301/commits/665c57c3d8528c040942e996ef91cbc11e656afe) to avoid turning off of any devices while testing. Only idle is checked because in this tests we only check if the shutdown tree was called, we dont need to check if it's execution is successful.
Because shutdown tree is so simple IMO tests for `ShutdownSingleHost` and `ShutdownHostsFromFile` bt nodes are sufficient.